### DISCUSSION TITLE
Requests hang if I use ignore_hosts or allow_hosts.

### DISCUSSION THREAD

**Discussion by akerl** [mitmproxy/mitmproxy#6781] [Q&A]
I'm trying to run a mitmproxy instance that modifies specific requests and ignores all other requests. I'm running mitmproxy via the official docker container, and I'm providing it 2 trusted certificates, one for the hostname of the proxy and one for the hostname I care about.

Here's how I'm invoking it for my test:

```
/usr/bin/docker run -ti --network=docker1 --ip 172.17.0.4 -v /mnt/frameproxy/cache:/opt/cache -v /mnt/frameproxy/tls:/opt/tls -v /mnt/frameproxy/scripts:/opt/scripts mitmproxy/mitmproxy mitmproxy --certs frame.akerl.app=/opt/tls/intercept_cert --certs frameproxy.a-rwx.org=/opt/tls/proxy_cert
```

With that running, I see the following:

1. With HTTPS_PROXY set to `http://frameproxy.a-rwx.org:8080` , `curl -v icanhazip.com` and `curl -v frame.akerl.app` both work.
2. With HTTPS_PROXY set to `https://frameproxy.a-rwx.org:8080` , `curl -v icanhazip.com` and `curl -v frame.akerl.app` both work.

If I modify the options to set allow_hosts to frame.akerl.app:443, `HTTPS_PROXY=http://frameproxy.a-rwx.org:8080` (non-HTTPS proxy) still works, but suddenly any requests with `HTTPS_PROXY=https://frameproxy.a-rwx.org:8080` just hang like this:

```
❯ curl -v https://frame.akerl.app/
* Uses proxy env variable HTTPS_PROXY == 'https://frameproxy.a-rwx.org:8080'
*   Trying 10.0.1.80:8080...
* Connected to frameproxy.a-rwx.org (10.0.1.80) port 8080 (#0)
* ALPN: offers http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384
* ALPN: server accepted http/1.1
* Proxy certificate:
*  subject: CN=frameproxy.a-rwx.org
*  start date: Apr  4 02:25:34 2024 GMT
*  expire date: Jul  3 02:25:33 2024 GMT
*  subjectAltName: host "frameproxy.a-rwx.org" matched cert's "frameproxy.a-rwx.org"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
* CONNECT tunnel: HTTP/1.1 negotiated
* allocate connect buffer
* Establish HTTP proxy tunnel to frame.akerl.app:443
> CONNECT frame.akerl.app:443 HTTP/1.1
> Host: frame.akerl.app:443
> User-Agent: curl/8.1.2
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection established
<
* CONNECT phase completed
* CONNECT tunnel established, response 200
* ALPN: offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
```

I never see any indication in the mitmproxy interface of the connection. the CONNECT is sending the matching hostname.

I get the exact same behavior if instead of allow_hosts, I set ignore_hosts to icanhazip.com:443: in both cases, setting the filter causes all requests to the proxy via HTTPS to stall.

What am I missing?

**Comment by rosydawn6** [comment]
I had something similar to this when doing testing with ignore_hosts. See if this thread has any applicability in your case: https://github.com/mitmproxy/mitmproxy/discussions/5638

@mhils suggested to vary the https version that curl was using and that provided insight into what was going on.
